table/sstable/table_file.go

```diff
@@ -170,19 +175,31 @@ func NewLocalFile(filename string, blockCache, indexCache *cache.Cache) (TableFi
 	return reader, nil
 }
 
-func (r *LocalFile) ReadBlock(offset, length int64) ([]byte, error) {
+func (r *LocalFile) readBlock(offset, length int64) (*block, error) {
 	if ptr := atomic.LoadPointer(&r.memReader); ptr != nil {
-		return (*InMemFile)(ptr).ReadBlock(offset, length)
+		return (*InMemFile)(ptr).readBlock(offset, length)
 	}
-	blockData, err := r.blockCache.GetOrCompute(blockCacheKey(r.fid, uint32(offset)), func() (interface{}, int64, error) {
-		data := make([]byte, length)
-		_, err := r.fd.ReadAt(data, offset)
-		return data, length, err
+	blk, err := r.blockCache.GetOrCompute(blockCacheKey(r.fid, uint32(offset)), func() (interface{}, int64, error) {
+		data := buffer.GetBuffer(int(length))
+		if _, err := r.fd.ReadAt(data, offset); err != nil {
+			buffer.PutBuffer(data)
+			return nil, 0, err
+		}
+		blk := &block{
+			offset:    int(offset),
+			data:      data,
+			reference: 1,
+		}
+		return blk, length, nil
 	})
 	if err != nil {
-		return nil, err
+		return &block{}, err
+	}
+	b := blk.(*block)
+	if ok := b.add(); !ok {
+		return &block{}, errors.Errorf("block is evicted")
```

Comment:
we can retry for this error, so the outer user don't need to handle this error.

Reply:
fixed